diff -ruN linux-4.4-orig/net/ipv4/tcp_ipv4.c linux-4.4/net/ipv4/tcp_ipv4.c
--- linux-4.4-orig/net/ipv4/tcp_ipv4.c	2016-04-01 01:45:43.155889438 -0700
+++ linux-4.4/net/ipv4/tcp_ipv4.c	2016-04-01 21:28:26.823678256 -0700
@@ -1619,6 +1619,7 @@
 		goto discard_and_relse;
 	}
 
+        if (!(skb->mark&0x10000000))
 	if (!xfrm4_policy_check(sk, XFRM_POLICY_IN, skb))
 		goto discard_and_relse;
 
@@ -1659,6 +1660,7 @@
 	return ret;
 
 no_tcp_socket:
+        if (!(skb->mark&0x10000000))
 	if (!xfrm4_policy_check(NULL, XFRM_POLICY_IN, skb))
 		goto discard_it;
 
@@ -1681,6 +1683,7 @@
 	goto discard_it;
 
 do_time_wait:
+        if (!(skb->mark&0x10000000))
 	if (!xfrm4_policy_check(NULL, XFRM_POLICY_IN, skb)) {
 		inet_twsk_put(inet_twsk(sk));
 		goto discard_it;
