Index: linux-source-2.6.22-2.6.22/net/bridge/br_input.c
===================================================================
--- linux-source-2.6.22-2.6.22.orig/net/bridge/br_input.c	2007-07-08 16:32:17.000000000 -0700
+++ linux-source-2.6.22-2.6.22/net/bridge/br_input.c	2007-10-16 22:01:47.000000000 -0700
@@ -131,6 +131,10 @@
 	if (!is_valid_ether_addr(eth_hdr(skb)->h_source))
 		goto drop;
 
+	if (p->state == BR_STATE_LEARNING ||
+            p->state == BR_STATE_FORWARDING)
+		br_fdb_update(p->br, p, eth_hdr(skb)->h_source);
+
 	if (unlikely(is_link_local(dest))) {
 		/* Pause frames shouldn't be passed up by driver anyway */
 		if (skb->protocol == htons(ETH_P_PAUSE))
