--- linux-2.6.16-ck11-dev/net/bridge/br_input.c	2006-03-19 21:53:29.000000000 -0800
+++ linux-2.6.16-ck11-prod/net/bridge/br_input.c	2006-06-15 17:09:16.000000000 -0700
@@ -117,6 +117,10 @@
 	if (!is_valid_ether_addr(eth_hdr(skb)->h_source))
 		goto err;
 
+	if (p->state == BR_STATE_LEARNING ||
+        p->state == BR_STATE_FORWARDING)
+		br_fdb_update(p->br, p, eth_hdr(skb)->h_source);
+
 	if (p->br->stp_enabled &&
 	    !memcmp(dest, bridge_ula, 5) &&
 	    !(dest[5] & 0xF0)) {
