--- linux-4.9/drivers/net/wireless/Kconfig.bak	2017-05-17 15:24:41.850389787 +0200
+++ linux-4.9/drivers/net/wireless/Kconfig	2017-05-17 15:25:02.922035293 +0200
@@ -28,16 +28,16 @@
 source "drivers/net/wireless/mediatek/Kconfig"
 source "drivers/net/wireless/ralink/Kconfig"
 source "drivers/net/wireless/realtek/Kconfig"
-if RALINK_DEVICE
-source "drivers/net/wireless/rt3352/rt2860v2_ap/Kconfig"
-source "drivers/net/wireless/rt3352/rt2860v2_sta/Kconfig"
-endif
-if SOC_MT7620_OPENWRT
-source "drivers/net/wireless/rt7620/rt2860v2_ap/Kconfig"
-source "drivers/net/wireless/rt7620/rt2860v2_sta/Kconfig"
-source "drivers/net/wireless/rt5592/Kconfig"
-source "drivers/net/wireless/rt7612/rlt_wifi/Kconfig"
-endif
+# if RALINK_DEVICE
+# source "drivers/net/wireless/rt3352/rt2860v2_ap/Kconfig"
+# source "drivers/net/wireless/rt3352/rt2860v2_sta/Kconfig"
+# endif
+# if SOC_MT7620_OPENWRT
+# source "drivers/net/wireless/rt7620/rt2860v2_ap/Kconfig"
+# source "drivers/net/wireless/rt7620/rt2860v2_sta/Kconfig"
+# source "drivers/net/wireless/rt5592/Kconfig"
+# source "drivers/net/wireless/rt7612/rlt_wifi/Kconfig"
+# endif
 source "drivers/net/wireless/rsi/Kconfig"
 source "drivers/net/wireless/st/Kconfig"
 source "drivers/net/wireless/ti/Kconfig"
