--- ./debian/config/amd64/defines	2026-02-17 18:06:14.650813955 +0530
+++ ./debian/config/amd64/defines	2026-02-17 18:09:25.080812793 +0530
@@ -1,5 +1,6 @@
 [base]
 featuresets:
+ untangle
  none
  rt
 kernel-arch: x86
--- ./debian/config/amd64/untangle/defines	1970-01-01 05:30:00.000000000 +0530
+++ ./debian/config/amd64/untangle/defines	2026-02-17 18:09:36.910812720 +0530
@@ -0,0 +1,6 @@
+[base]
+flavours:
+ amd64
+
+[amd64_image]
+provides: linux-image-untangle, linux-image-6.1-untangle, linux-image-6.1-untangle-amd64
