diff -ruN linux-4.4.6.orig/net/bridge/br_input.c linux-4.4.6/net/bridge/br_input.c
--- linux-4.4.6.orig/net/bridge/br_input.c	2016-03-17 23:31:50.337629464 -0700
+++ linux-4.4.6/net/bridge/br_input.c	2016-03-17 23:38:22.849337249 -0700
@@ -248,6 +248,14 @@
 
 	p = br_port_get_rcu(skb->dev);
 
+	if (p->state == BR_STATE_LEARNING ||
+            p->state == BR_STATE_FORWARDING) {
+               u16 vid = 0;
+               br_vlan_get_tag(skb, &vid);
+	       br_fdb_update(p->br, p, eth_hdr(skb)->h_source, vid, false);
+        }
+
+
 	if (unlikely(is_link_local_ether_addr(dest))) {
 		u16 fwd_mask = p->br->group_fwd_mask_required;
 
