diff -ruN linux-4.4-orig/net/netfilter/xt_socket.c linux-4.4/net/netfilter/xt_socket.c
--- linux-4.4-orig/net/netfilter/xt_socket.c	2016-04-01 01:45:43.147889377 -0700
+++ linux-4.4/net/netfilter/xt_socket.c	2016-04-01 02:02:23.227422368 -0700
@@ -238,8 +238,14 @@
 		if (wildcard || !transparent)
 			sk = NULL;
 	}
-
-	return sk != NULL;
+    /* Untangle hack - or the socket mark to the packet if found
+         * We do this because there is no conntrack, and it is necessary
+         * to restore the QoS/bandwidth control mark on packets
+         * returning to nonlocally bound sockets */
+    if (sk != NULL) {
+        skb->mark |= sk->sk_mark;
+    }
+    return sk != NULL;
 }
 
 static bool
