--- /home/seb/work/untangle/dd-wrt/dd-wrt-github/src/linux/universal/linux-4.4/.config_northstar_smp	2016-10-13 13:06:38.278963841 +0200
+++ linux-4.4/.config_northstar_smp	2016-11-23 12:17:38.122048396 +0100
@@ -211,7 +211,7 @@
 # CONFIG_MODVERSIONS is not set
 # CONFIG_MODULE_SRCVERSION_ALL is not set
 # CONFIG_MODULE_SIG is not set
-CONFIG_MODULE_STRIPPED=y
+# CONFIG_MODULE_STRIPPED is not set
 # CONFIG_MODULE_COMPRESS is not set
 CONFIG_BLOCK=y
 CONFIG_LBDAF=y
@@ -2085,7 +2085,9 @@
 #
 CONFIG_DCACHE_WORD_ACCESS=y
 # CONFIG_EXT2_FS is not set
+# CONFIG_EXT2_FS_XATTR is not set
 # CONFIG_EXT3_FS is not set
+# CONFIG_EXT3_FS_SECURITY is not set
 CONFIG_EXT4_FS=m
 CONFIG_EXT4_USE_FOR_EXT2=y
 CONFIG_EXT4_FS_POSIX_ACL=y
@@ -1307,7 +1307,7 @@
 #
 # Broadcom HND network devices
 #
-CONFIG_HND=y
+CONFIG_HND=m
 CONFIG_ET=m
 CONFIG_ET_47XX=y
 # CONFIG_ET_ALL_PASSIVE_ON is not set
--- /home/seb/work/untangle/dd-wrt/dd-wrt-github/src/linux/universal/linux-4.4/.config_northstar_smp	2016-10-13 13:06:38.278963841 +0200
+++ linux-4.4/.config_northstar_smp	2016-11-23 15:28:40.650244792 +0100
@@ -1630,7 +1630,7 @@
 # CONFIG_TCG_TPM is not set
 CONFIG_DEVPORT=y
 # CONFIG_XILLYBUS is not set
-CONFIG_RTL8365MB=m
+# CONFIG_RTL8365MB is not set
 
 #
 # I2C support
