--- /home/seb/work/untangle/dd-wrt/dd-wrt-github/src/linux/universal/linux-4.4/.config_northstar_smp	2016-10-13 13:06:38.278963841 +0200
+++ linux-4.4/.config_northstar_smp	2016-11-23 12:17:38.122048396 +0100
@@ -211,7 +211,7 @@
 # CONFIG_MODVERSIONS is not set
 # CONFIG_MODULE_SRCVERSION_ALL is not set
 # CONFIG_MODULE_SIG is not set
-CONFIG_MODULE_STRIPPED=y
+# CONFIG_MODULE_STRIPPED is not set
 # CONFIG_MODULE_COMPRESS is not set
 CONFIG_BLOCK=y
 CONFIG_LBDAF=y
@@ -1315,8 +1315,8 @@
 # CONFIG_BCM_CTF is not set
 # CONFIG_WL_EMULATOR is not set
 # CONFIG_BCM57XX is not set
-CONFIG_DHDAP=m
-CONFIG_WL=m
+# CONFIG_DHDAP is not set
+# CONFIG_WL is not set
 # CONFIG_WAPI is not set
 # CONFIG_WL_USE_AP is not set
 # CONFIG_WL_USE_AP_SDSTD is not set
@@ -1630,7 +1630,7 @@
 # CONFIG_TCG_TPM is not set
 CONFIG_DEVPORT=y
 # CONFIG_XILLYBUS is not set
-CONFIG_RTL8365MB=m
+# CONFIG_RTL8365MB is not set
 
 #
 # I2C support
@@ -2085,7 +2085,9 @@
 #
 CONFIG_DCACHE_WORD_ACCESS=y
 # CONFIG_EXT2_FS is not set
+# CONFIG_EXT2_FS_XATTR is not set
 # CONFIG_EXT3_FS is not set
+# CONFIG_EXT3_FS_SECURITY is not set
 CONFIG_EXT4_FS=m
 CONFIG_EXT4_USE_FOR_EXT2=y
 CONFIG_EXT4_FS_POSIX_ACL=y
