--- linux-4.4/.config_northstar_smp    2016-10-13 13:06:38.278963841 +0200
+++ linux-4.4/.config_northstar_smp     2016-11-23 12:17:38.122048396 +0100
@@ -1315,8 +1315,8 @@
 # CONFIG_BCM_CTF is not set
 # CONFIG_WL_EMULATOR is not set
 # CONFIG_BCM57XX is not set
-CONFIG_DHDAP=m
-CONFIG_WL=m
+# CONFIG_DHDAP is not set
+# CONFIG_WL is not set
 # CONFIG_WAPI is not set
 # CONFIG_WL_USE_AP is not set
 # CONFIG_WL_USE_AP_SDSTD is not set
