--- linux-4.4/.config_northstar_smp    2016-10-13 13:06:38.278963841 +0200
+++ linux-4.4/.config_northstar_smp     2016-11-23 12:17:38.122048396 +0100
@@ -2085,7 +2085,9 @@
 #
 CONFIG_DCACHE_WORD_ACCESS=y
 # CONFIG_EXT2_FS is not set
+# CONFIG_EXT2_FS_XATTR is not set
 # CONFIG_EXT3_FS is not set
+# CONFIG_EXT3_FS_SECURITY is not set
 CONFIG_EXT4_FS=m
 CONFIG_EXT4_USE_FOR_EXT2=y
 CONFIG_EXT4_FS_POSIX_ACL=y
