diff -ruN linux-4.4.6.orig/net/bridge/br_nf_core.c linux-4.4.6/net/bridge/br_nf_core.c
--- linux-4.4.6.orig/net/bridge/br_nf_core.c	2016-03-17 23:31:50.337629464 -0700
+++ linux-4.4.6/net/bridge/br_nf_core.c	2016-03-27 21:12:21.115268911 -0700
@@ -19,12 +19,24 @@
 #include <linux/in_route.h>
 #include <linux/inetdevice.h>
 #include <net/route.h>
+#include <net/arp.h>
 
 #include "br_private.h"
 #ifdef CONFIG_SYSCTL
 #include <linux/sysctl.h>
 #endif
 
+static inline struct nf_bridge_info *nf_bridge_alloc(struct sk_buff *skb)
+{
+        skb->nf_bridge = kzalloc(sizeof(struct nf_bridge_info), GFP_ATOMIC);
+
+            if (likely(skb->nf_bridge))
+                        atomic_set(&(skb->nf_bridge->use), 1);
+
+                return skb->nf_bridge;
+
+}
+
 static void fake_update_pmtu(struct dst_entry *dst, struct sock *sk,
 			     struct sk_buff *skb, u32 mtu)
 {
@@ -93,3 +105,81 @@
 {
 	dst_entries_destroy(&fake_dst_ops);
 }
+
+
+struct net_bridge_fdb_entry *br_fdb_get(struct net_bridge *br,
+					unsigned char *addr)
+{
+	struct net_bridge_fdb_entry *fdb;
+
+	rcu_read_lock();
+	fdb = __br_fdb_get(br, addr, 0 /* vid ??? FIXME */ );
+	rcu_read_unlock();
+	return fdb;
+}
+
+static void ngfw_arp_send(struct net_device *out_dev,
+			  __be32 dest_ip, __be32 src_ip)
+{
+	struct sk_buff *skb;
+
+	skb = arp_create(ARPOP_REQUEST, ETH_P_ARP, dest_ip, out_dev, src_ip,
+			 NULL, out_dev->dev_addr, NULL);
+
+	if (!skb) {
+		return;
+	}
+
+	arp_xmit(skb);
+}
+
+void get_outdev(struct sk_buff* skb, struct net_device *out, struct net_device *m_out)
+{
+	struct neighbour *neigh;
+	struct nf_bridge_info *nf_bridge;
+	unsigned char *dest;
+	struct net_bridge_fdb_entry *fdb;
+	struct dst_entry *dst = skb_dst(skb);
+	struct rtable *rt = (struct rtable *)dst;
+	struct net_device *dev = dst->dev;
+	u32 src_ip = 0;
+    u32 nexthop;
+
+	nf_bridge = skb->nf_bridge;
+	if ((nf_bridge && nf_bridge->physoutdev) ||
+	    !out || out->netdev_ops->ndo_start_xmit != br_dev_xmit)
+		return;
+
+	if (!nf_bridge && !(nf_bridge = nf_bridge_alloc(skb)))
+		return;
+	nf_bridge->physoutdev = out;
+
+	nexthop = (__force u32) rt_nexthop(rt, ip_hdr(skb)->daddr);
+	neigh = __ipv4_neigh_lookup_noref(dev, nexthop);
+	if (unlikely(!neigh)){
+        neigh = __neigh_create(&arp_tbl, &nexthop, dev, false);
+        if (dev)
+            src_ip = inet_select_addr(dev, nexthop, RT_SCOPE_LINK);
+        ngfw_arp_send(m_out,nexthop,src_ip);
+		return;
+    }
+	if (neigh_event_send(neigh, NULL)) {
+		return;
+    }
+
+	dest = neigh->ha;
+	if (!is_multicast_ether_addr(dest)) {
+        fdb = br_fdb_get(netdev_priv(out), dest);
+
+        if (fdb) {
+            nf_bridge->physoutdev = fdb->dst->dev;
+        }
+        else {
+            if (dev)
+                src_ip = inet_select_addr(dev, nexthop, RT_SCOPE_LINK);
+            ngfw_arp_send(m_out,nexthop,src_ip);
+        }
+    }
+}
+
+EXPORT_SYMBOL(get_outdev);
diff -ruN linux-4.4.6.orig/net/netfilter/xt_physdev.c linux-4.4.6/net/netfilter/xt_physdev.c
--- linux-4.4.6.orig/net/netfilter/xt_physdev.c	2016-03-17 23:31:51.113636812 -0700
+++ linux-4.4.6/net/netfilter/xt_physdev.c	2016-03-27 21:13:33.067841475 -0700
@@ -15,6 +15,8 @@
 #include <linux/netfilter/x_tables.h>
 #include <net/netfilter/br_netfilter.h>
 
+void get_outdev(const struct sk_buff* skb, const struct net_device *out, struct net_device *m_dev);
+
 MODULE_LICENSE("GPL");
 MODULE_AUTHOR("Bart De Schuymer <bdschuym@pandora.be>");
 MODULE_DESCRIPTION("Xtables: Bridge physical device match");
@@ -29,6 +31,8 @@
 	const struct net_device *physdev;
 	unsigned long ret;
 	const char *indev, *outdev;
+    struct nf_bridge_info *nf_bridge;
+    struct net_device  *m_dev;
 
 	/* Not a bridged IP packet or no info available yet:
 	 * LOCAL_OUT/mangle and LOCAL_OUT/nat don't know if
@@ -41,16 +45,10 @@
 		if ((info->bitmask & XT_PHYSDEV_OP_ISIN) &&
 		    !(info->invert & XT_PHYSDEV_OP_ISIN))
 			return false;
-		if ((info->bitmask & XT_PHYSDEV_OP_ISOUT) &&
-		    !(info->invert & XT_PHYSDEV_OP_ISOUT))
-			return false;
 		if ((info->bitmask & XT_PHYSDEV_OP_IN) &&
 		    !(info->invert & XT_PHYSDEV_OP_IN))
 			return false;
-		if ((info->bitmask & XT_PHYSDEV_OP_OUT) &&
-		    !(info->invert & XT_PHYSDEV_OP_OUT))
-			return false;
-		return true;
+        goto match_outdev;
 	}
 
 	physdev = nf_bridge_get_physoutdev(skb);
@@ -65,9 +63,7 @@
 	indev = physdev ? physdev->name : NULL;
 
 	if ((info->bitmask & XT_PHYSDEV_OP_ISIN &&
-	    (!indev ^ !!(info->invert & XT_PHYSDEV_OP_ISIN))) ||
-	    (info->bitmask & XT_PHYSDEV_OP_ISOUT &&
-	    (!outdev ^ !!(info->invert & XT_PHYSDEV_OP_ISOUT))))
+	    (!indev ^ !!(info->invert & XT_PHYSDEV_OP_ISIN))))
 		return false;
 
 	if (!(info->bitmask & XT_PHYSDEV_OP_IN))
@@ -82,8 +78,29 @@ physdev_mt(const struct sk_buff *skb, struct xt_action_param *par)
 	}
 
 match_outdev:
-	if (!(info->bitmask & XT_PHYSDEV_OP_OUT))
+    if (!(info->bitmask & (XT_PHYSDEV_OP_ISOUT | XT_PHYSDEV_OP_OUT)))
+		return true;
+
+	m_dev = dev_get_by_name(&init_net, info->physoutdev );
+
+	get_outdev(skb, par->out, m_dev);
+	dev_put(m_dev);
+	if (!(nf_bridge = skb->nf_bridge)) {
+		if ((info->bitmask & XT_PHYSDEV_OP_ISOUT) &&
+		    !(info->invert & XT_PHYSDEV_OP_ISOUT))
+			return false;
+		if ((info->bitmask & XT_PHYSDEV_OP_OUT) &&
+		    !(info->invert & XT_PHYSDEV_OP_OUT))
+			return false;
 		return true;
+	}
+
+	if (info->bitmask & XT_PHYSDEV_OP_ISOUT &&
+	    (!nf_bridge->physoutdev ^ !!(info->invert & XT_PHYSDEV_OP_ISOUT)))
+		return false;
+
+	physdev = nf_bridge_get_physoutdev(skb);
+	outdev = physdev ? physdev->name : NULL;
 
 	if (!outdev)
 		return false;
@@ -102,17 +119,7 @@ static int physdev_mt_check(const struct xt_mtchk_param *par)
 	if (!(info->bitmask & XT_PHYSDEV_OP_MASK) ||
 	    info->bitmask & ~XT_PHYSDEV_OP_MASK)
 		return -EINVAL;
-	if (info->bitmask & XT_PHYSDEV_OP_OUT &&
-	    (!(info->bitmask & XT_PHYSDEV_OP_BRIDGED) ||
-	     info->invert & XT_PHYSDEV_OP_BRIDGED) &&
-	    par->hook_mask & ((1 << NF_INET_LOCAL_OUT) |
-	    (1 << NF_INET_FORWARD) | (1 << NF_INET_POST_ROUTING))) {
-		pr_info("using --physdev-out in the OUTPUT, FORWARD and "
-			"POSTROUTING chains for non-bridged traffic is not "
-			"supported anymore.\n");
-		if (par->hook_mask & (1 << NF_INET_LOCAL_OUT))
-			return -EINVAL;
-	}
+
 	return 0;
 }
 
