diff -ruN linux-4.4-orig/include/net/netns/ipv4.h linux-4.4/include/net/netns/ipv4.h
--- linux-4.4-orig/include/net/netns/ipv4.h	2016-04-01 01:45:42.495884499 -0700
+++ linux-4.4/include/net/netns/ipv4.h	2016-04-01 21:33:34.306150904 -0700
@@ -61,6 +61,7 @@
 	struct xt_table		*iptable_filter;
 	struct xt_table		*iptable_mangle;
 	struct xt_table		*iptable_raw;
+	struct xt_table		*iptable_tune;
 	struct xt_table		*arptable_filter;
 #ifdef CONFIG_SECURITY
 	struct xt_table		*iptable_security;
diff -ruN linux-4.4-orig/include/uapi/linux/netfilter_ipv4.h linux-4.4/include/uapi/linux/netfilter_ipv4.h
--- linux-4.4-orig/include/uapi/linux/netfilter_ipv4.h	2016-04-01 01:45:42.479884380 -0700
+++ linux-4.4/include/uapi/linux/netfilter_ipv4.h	2016-04-01 21:33:34.310150936 -0700
@@ -67,7 +67,8 @@
 	NF_IP_PRI_NAT_SRC = 100,
 	NF_IP_PRI_SELINUX_LAST = 225,
 	NF_IP_PRI_CONNTRACK_HELPER = 300,
-	NF_IP_PRI_CONNTRACK_CONFIRM = INT_MAX,
+	NF_IP_PRI_CONNTRACK_CONFIRM = INT_MAX - 10,
+	NF_IP_PRI_TUNE = INT_MAX - 5,
 	NF_IP_PRI_LAST = INT_MAX,
 };
 
diff -ruN linux-4.4-orig/net/ipv4/netfilter/iptable_tune.c linux-4.4/net/ipv4/netfilter/iptable_tune.c
--- linux-4.4-orig/net/ipv4/netfilter/iptable_tune.c	1969-12-31 16:00:00.000000000 -0800
+++ linux-4.4/net/ipv4/netfilter/iptable_tune.c	2016-04-01 21:33:34.310150936 -0700
@@ -0,0 +1,94 @@
+/* 
+ * 
+ * (C) 2015 David Han <dhan@untangle.com>
+ *
+ * Port old version tune table to kernel 3.16 version
+ */
+
+#include <linux/module.h>
+#include <linux/netfilter.h>
+#include <linux/netfilter_ipv4.h>
+#include <linux/netfilter_ipv4/ip_tables.h>
+#include <linux/ip.h>
+#include <net/ip.h>
+
+#define TUNE_VALID_HOOKS ((1 << NF_INET_POST_ROUTING) | (1 << NF_INET_LOCAL_OUT))
+
+static struct nf_hook_ops *tune_ops __read_mostly;
+
+static const struct xt_table packet_tuner = {
+	.name		= "tune",
+	.valid_hooks	= TUNE_VALID_HOOKS,
+	.me		= THIS_MODULE,
+	.af		= NFPROTO_IPV4,
+	.priority = NF_IP_PRI_TUNE,
+};
+
+static unsigned int
+iptable_tune_hook(void *priv, struct sk_buff *skb,
+		    const struct nf_hook_state *state)
+{
+	if (state->hook == NF_INET_LOCAL_OUT &&
+	    (skb->len < sizeof(struct iphdr) ||
+	     ip_hdrlen(skb) < sizeof(struct iphdr)))
+		/* root is playing with raw sockets. */
+		return NF_ACCEPT;
+
+	return ipt_do_table(skb, state, state->net->ipv4.iptable_tune);
+}
+
+
+
+static int __net_init iptable_tune_net_init(struct net *net)
+{
+	struct ipt_replace *repl;
+
+	repl = ipt_alloc_initial_table(&packet_tuner);
+	if (repl == NULL)
+		return -ENOMEM;
+
+	net->ipv4.iptable_tune = ipt_register_table(net, &packet_tuner, repl);
+	kfree(repl);
+	return PTR_ERR_OR_ZERO(net->ipv4.iptable_tune);
+}
+
+
+
+static void __net_exit iptable_tune_net_exit(struct net *net)
+{
+	ipt_unregister_table(net, net->ipv4.iptable_tune);
+}
+
+static struct pernet_operations iptable_tune_net_ops = {
+	.init	= iptable_tune_net_init,
+	.exit	= iptable_tune_net_exit,
+};
+
+static int __init iptable_tune_init(void)
+{
+	int ret;
+
+	ret = register_pernet_subsys(&iptable_tune_net_ops);
+	if (ret < 0)
+		return ret;
+
+	/* Register hooks */
+	tune_ops = xt_hook_link(&packet_tuner, iptable_tune_hook);
+	if (IS_ERR(tune_ops)) {
+		ret = PTR_ERR(tune_ops);
+		unregister_pernet_subsys(&iptable_tune_net_ops);
+	}
+
+	return ret;
+}
+
+static void __exit iptable_tune_exit(void)
+{
+	xt_hook_unlink(&packet_tuner, tune_ops);
+	unregister_pernet_subsys(&iptable_tune_net_ops);
+}
+
+module_init(iptable_tune_init);
+module_exit(iptable_tune_exit);
+
+MODULE_LICENSE("GPL");
diff -ruN linux-4.4-orig/net/ipv4/netfilter/Kconfig linux-4.4/net/ipv4/netfilter/Kconfig
--- linux-4.4-orig/net/ipv4/netfilter/Kconfig	2016-04-01 01:45:43.155889438 -0700
+++ linux-4.4/net/ipv4/netfilter/Kconfig	2016-04-01 21:33:34.310150936 -0700
@@ -443,3 +443,15 @@
 
 endmenu
 
+# tune + specific targets
+config IP_NF_TUNE
+	tristate  'tune table support (adds a chain after POSTROUTING nat)'
+	depends on IP_NF_IPTABLES
+	help
+	  This option adds a `tune' table to iptables. This table is the very
+	  last in the netfilter framework and hooks in at the POSTROUTING
+	  chain.
+	
+	  If you want to compile it as a module, say M here and read
+	  <file:Documentation/modules.txt>.  If unsure, say `N'.
+
diff -ruN linux-4.4-orig/net/ipv4/netfilter/Makefile linux-4.4/net/ipv4/netfilter/Makefile
--- linux-4.4-orig/net/ipv4/netfilter/Makefile	2016-04-01 01:45:43.155889438 -0700
+++ linux-4.4/net/ipv4/netfilter/Makefile	2016-04-01 21:33:34.310150936 -0700
@@ -54,6 +54,7 @@
 obj-$(CONFIG_IP_NF_MANGLE) += iptable_mangle.o
 obj-$(CONFIG_IP_NF_NAT) += iptable_nat.o
 obj-$(CONFIG_IP_NF_RAW) += iptable_raw.o
+obj-$(CONFIG_IP_NF_TUNE) += iptable_tune.o
 obj-$(CONFIG_IP_NF_SECURITY) += iptable_security.o
 
 # matches
