From 020e1e82ef673a74227dfdca57b3a6f89473928e Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@freescale.com>
Date: Thu, 3 Nov 2011 14:14:09 +0800
Subject: ENGR00161314-3 mx6q usb-host: add hsic support

Header file

Signed-off-by: Peter Chen <peter.chen@freescale.com>
---
 include/linux/fsl_devices.h |    2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/linux/fsl_devices.h b/include/linux/fsl_devices.h
index 20cce21..7508142 100644
--- a/include/linux/fsl_devices.h
+++ b/include/linux/fsl_devices.h
@@ -65,6 +65,7 @@ enum fsl_usb2_phy_modes {
 	FSL_USB2_PHY_UTMI,
 	FSL_USB2_PHY_UTMI_WIDE,
 	FSL_USB2_PHY_SERIAL,
+	FSL_USB2_PHY_HSIC,
 };
 
 enum usb_wakeup_event {
@@ -121,6 +122,7 @@ struct fsl_usb2_platform_data {
 	void (*platform_driver_vbus)(bool on); /* for vbus shutdown/open */
 	enum usb_wakeup_event (*is_wakeup_event)(struct fsl_usb2_platform_data *);
 	void (*wakeup_handler)(struct fsl_usb2_platform_data *);
+	void (*hsic_post_ops)(void);
 
 	struct fsl_usb2_wakeup_platform_data *wakeup_pdata;
 	struct platform_device *pdev;
-- 
1.7.9.5

