From 36d57d4587b1e8790e9067964154c00746607b3e Mon Sep 17 00:00:00 2001
From: Wayne Zou <b36644@freescale.com>
Date: Fri, 16 Dec 2011 10:20:48 +0800
Subject: mxcfb: add BACKLIGHT_RAW type for kernel upgrade

Signed-off-by: Wayne Zou <b36644@freescale.com>
---
 drivers/video/mxc/mxcfb_hx8369_wvga.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/video/mxc/mxcfb_hx8369_wvga.c b/drivers/video/mxc/mxcfb_hx8369_wvga.c
index 2d89aa0..d8771fd 100644
--- a/drivers/video/mxc/mxcfb_hx8369_wvga.c
+++ b/drivers/video/mxc/mxcfb_hx8369_wvga.c
@@ -374,6 +374,7 @@ static int mipid_init_backlight(struct mipi_dsi_info *mipi_dsi)
 	}
 	memset(&props, 0, sizeof(struct backlight_properties));
 	props.max_brightness = HX8369BL_MAX_BRIGHT;
+	props.type = BACKLIGHT_RAW;
 	bl = backlight_device_register("mipid-bl", &mipi_dsi->pdev->dev,
 		mipi_dsi, &mipid_lcd_bl_ops, &props);
 	if (IS_ERR(bl)) {
-- 
1.7.9.5

