From 913fcf5d527f71de7909e0d921a001a652292e16 Mon Sep 17 00:00:00 2001
From: Jason Chen <jason.chen@linaro.org>
Date: Tue, 27 Dec 2011 16:12:07 +0800
Subject: mxc: hdmi: fix potention deadlock issue

Signed-off-by: Jason Chen <jason.chen@linaro.org>
---
 drivers/video/mxc_hdmi.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/video/mxc_hdmi.c b/drivers/video/mxc_hdmi.c
index 3c1f8fd..9e83df0 100644
--- a/drivers/video/mxc_hdmi.c
+++ b/drivers/video/mxc_hdmi.c
@@ -2033,12 +2033,12 @@ static void mxc_hdmi_fb_registered(struct mxc_hdmi *hdmi)
 	if (hdmi->fb_reg)
 		return;
 
+	clk_enable(hdmi->hdmi_iahb_clk);
+
 	spin_lock_irqsave(&hdmi->irq_lock, flags);
 
 	dev_dbg(&hdmi->pdev->dev, "%s\n", __func__);
 
-	clk_enable(hdmi->hdmi_iahb_clk);
-
 	hdmi_writeb(HDMI_PHY_I2CM_INT_ADDR_DONE_POL,
 		    HDMI_PHY_I2CM_INT_ADDR);
 
@@ -2057,9 +2057,9 @@ static void mxc_hdmi_fb_registered(struct mxc_hdmi *hdmi)
 
 	hdmi->fb_reg = true;
 
-	clk_disable(hdmi->hdmi_iahb_clk);
-
 	spin_unlock_irqrestore(&hdmi->irq_lock, flags);
+
+	clk_disable(hdmi->hdmi_iahb_clk);
 }
 
 static int mxc_hdmi_fb_event(struct notifier_block *nb,
-- 
1.7.9.5

