From 0fa3000a8e9d43a2c2b9f9aa8b87a105f519ca2d Mon Sep 17 00:00:00 2001
From: Lionel Xu <Lionel.Xu@freescale.com>
Date: Wed, 13 Jun 2012 14:00:07 +0800
Subject: ENGR00163374 ESAI: ESAI driver does not support mono playback

To remove mono playback support for ESAI.

Signed-off-by: Lionel Xu <Lionel.Xu@freescale.com>
---
 sound/soc/imx/imx-esai.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/sound/soc/imx/imx-esai.c b/sound/soc/imx/imx-esai.c
index 67eedfa..daa3a8a 100644
--- a/sound/soc/imx/imx-esai.c
+++ b/sound/soc/imx/imx-esai.c
@@ -569,7 +569,7 @@ static struct snd_soc_dai_driver imx_esai_dai = {
 	 .suspend = imx_esai_suspend,
 	 .resume = imx_esai_resume,
 	 .playback = {
-		      .channels_min = 1,
+		      .channels_min = 2,
 		      .channels_max = 12,
 		      .rates = IMX_ESAI_RATES,
 		      .formats = IMX_ESAI_FORMATS,
-- 
1.7.9.5

