From 3011a87407953fe5f03f1b78c0401a6dd1beeb56 Mon Sep 17 00:00:00 2001
From: Valentin Raevsky <valentin@compulab.co.il>
Date: Tue, 2 Jul 2013 13:14:35 +0300
Subject: ARM: i.MX6: cm-fx6: suppress annoying debug messages

---
 arch/arm/plat-mxc/cpufreq.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/arm/plat-mxc/cpufreq.c b/arch/arm/plat-mxc/cpufreq.c
index 47a70d9..1a42f9e 100755
--- a/arch/arm/plat-mxc/cpufreq.c
+++ b/arch/arm/plat-mxc/cpufreq.c
@@ -123,15 +123,15 @@ int set_cpu_freq(int freq)
 		ret = regulator_set_voltage(cpu_regulator, gp_volt,
 					    gp_volt);
 		if (ret < 0) {
-			printk(KERN_ERR "COULD NOT SET GP VOLTAGE!!!!\n");
+			/* printk(KERN_ERR "COULD NOT SET GP VOLTAGE!!!!\n"); */
 			goto err5;
 		}
 		if (!IS_ERR(soc_regulator)) {
 			ret = regulator_set_voltage(soc_regulator, soc_volt,
 							soc_volt);
 			if (ret < 0) {
-				printk(KERN_ERR
-					"COULD NOT SET SOC VOLTAGE BACK!!!!\n");
+				/* printk(KERN_ERR
+					"COULD NOT SET SOC VOLTAGE BACK!!!!\n"); */
 				goto err6;
 			}
 		}
-- 
1.7.9.5

