diff -ruN linux-4.4-orig/net/bridge/br_input.c linux-4.4/net/bridge/br_input.c
--- linux-4.4-orig/net/bridge/br_input.c	2016-04-01 01:45:43.127889228 -0700
+++ linux-4.4/net/bridge/br_input.c	2016-04-01 21:26:40.122822555 -0700
@@ -254,6 +254,14 @@
 
 	p = br_port_get_rcu(skb->dev);
 
+	if (p->state == BR_STATE_LEARNING ||
+            p->state == BR_STATE_FORWARDING) {
+               u16 vid = 0;
+               br_vlan_get_tag(skb, &vid);
+	       br_fdb_update(p->br, p, eth_hdr(skb)->h_source, vid, false);
+        }
+
+
 	if (unlikely(is_link_local_ether_addr(dest))) {
 		u16 fwd_mask = p->br->group_fwd_mask_required;
 
