--- tmp/net/ipv4/tcp_ipv4.c	2014-12-17 17:06:24.262831341 -0700
+++ linux-3.2.51/net/ipv4/tcp_ipv4.c	2014-12-19 01:10:47.684485475 -0700
@@ -1710,6 +1710,7 @@
 		goto discard_and_relse;
 	}
 
+        if (skb->mark)
 	if (!xfrm4_policy_check(sk, XFRM_POLICY_IN, skb))
 		goto discard_and_relse;
 	nf_reset(skb);
@@ -1746,6 +1747,7 @@
 	return ret;
 
 no_tcp_socket:
+        if (skb->mark)
 	if (!xfrm4_policy_check(NULL, XFRM_POLICY_IN, skb))
 		goto discard_it;
 
@@ -1766,6 +1768,7 @@
 	goto discard_it;
 
 do_time_wait:
+        if (skb->mark)
 	if (!xfrm4_policy_check(NULL, XFRM_POLICY_IN, skb)) {
 		inet_twsk_put(inet_twsk(sk));
 		goto discard_it;
