diff -urN linux-3.10.49.orig/include/uapi/linux/netfilter/nfnetlink_queue.h linux-3.10.49-014/include/uapi/linux/netfilter/nfnetlink_queue.h
--- linux-3.10.49.orig/include/uapi/linux/netfilter/nfnetlink_queue.h	2014-07-28 19:35:34.000000000 -0700
+++ linux-3.10.49-014/include/uapi/linux/netfilter/nfnetlink_queue.h	2014-07-28 20:02:45.000000000 -0700
@@ -47,6 +47,12 @@
 	NFQA_CAP_LEN,			/* __u32 length of captured packet */
 	NFQA_SKB_INFO,			/* __u32 skb meta information */
 
+#ifdef CONFIG_NFNETLINK_COPY_UNTANGLE_MODE
+	NFQA_CTINFO,
+	NFQA_CT_DIR_ORIGINAL,
+	NFQA_CT_DIR_REPLY,
+#endif
+
 	__NFQA_MAX
 };
 #define NFQA_MAX (__NFQA_MAX - 1)
@@ -77,6 +83,12 @@
 	NFQNL_COPY_PACKET,
 };
 
+#ifdef CONFIG_NFNETLINK_COPY_UNTANGLE_MODE
+/* ORed with  nfqnl_config_mode to indicate that */
+/* conntrack info should be added to each queued packet */
+#define NFQNL_COPY_UNTANGLE_MODE 0x10
+#endif
+
 struct nfqnl_msg_config_params {
 	__be32		copy_range;
 	__u8	copy_mode;	/* enum nfqnl_config_mode */
diff -urN linux-3.10.49.orig/net/netfilter/Kconfig linux-3.10.49-014/net/netfilter/Kconfig
--- linux-3.10.49.orig/net/netfilter/Kconfig	2014-07-28 19:35:34.000000000 -0700
+++ linux-3.10.49-014/net/netfilter/Kconfig	2014-07-28 20:02:45.000000000 -0700
@@ -19,7 +19,15 @@
 	help
 	  If this option is enabled, the kernel will include support
 	  for queueing packets via NFNETLINK.
-	  
+
+config NFNETLINK_COPY_UNTANGLE_MODE
+	bool "NFNETLINK untangle copy mode"
+	depends on NETFILTER_NETLINK
+	help
+	  If this option is enabled, the kernel will include support
+	  for appending conntrack info to queueed packets. to use this
+	  requires the use of untnagle's libnetfilter_queue.
+
 config NETFILTER_NETLINK_LOG
 	tristate "Netfilter LOG over NFNETLINK interface"
 	default m if NETFILTER_ADVANCED=n
diff -urN linux-3.10.49.orig/net/netfilter/nfnetlink_queue_core.c linux-3.10.49-014/net/netfilter/nfnetlink_queue_core.c
--- linux-3.10.49.orig/net/netfilter/nfnetlink_queue_core.c	2014-07-28 19:35:34.000000000 -0700
+++ linux-3.10.49-014/net/netfilter/nfnetlink_queue_core.c	2014-07-28 20:04:28.000000000 -0700
@@ -33,6 +33,11 @@
 #include <net/netns/generic.h>
 #include <net/netfilter/nfnetlink_queue.h>
 
+#ifdef CONFIG_NFNETLINK_COPY_UNTANGLE_MODE
+#include <net/netfilter/nf_conntrack.h>
+#include <net/netfilter/nf_conntrack_core.h>
+#endif
+
 #include <linux/atomic.h>
 
 #ifdef CONFIG_BRIDGE_NETFILTER
@@ -63,6 +68,10 @@
 	unsigned int	queue_total;
 	unsigned int	id_sequence;		/* 'sequence' of pkt ids */
 	struct list_head queue_list;		/* packets in queue */
+
+#ifdef CONFIG_NFNETLINK_COPY_UNTANGLE_MODE
+	int untangle_mode;			/* append conntrack info to queued packets*/
+#endif
 };
 
 typedef int (*nfqnl_cmpfn)(struct nf_queue_entry *, unsigned long);
@@ -119,6 +128,9 @@
 		goto out_unlock;
 	}
 
+#ifdef CONFIG_NFNETLINK_COPY_UNTANGLE_MODE
+	inst->untangle_mode = 0;
+#endif
 	inst->queue_num = queue_num;
 	inst->peer_portid = portid;
 	inst->queue_maxlen = NFQNL_QMAX_DEFAULT;
@@ -321,6 +333,14 @@
 
 	outdev = entry->outdev;
 
+#ifdef CONFIG_NFNETLINK_COPY_UNTANGLE_MODE
+	if( queue->untangle_mode){
+	        size += nla_total_size(sizeof(enum ip_conntrack_info))
+		    + nla_total_size(sizeof(struct nf_conntrack_tuple))  /* dir original */
+		    + nla_total_size(sizeof(struct nf_conntrack_tuple));  /* dir reply */
+	}
+#endif
+    
 	switch ((enum nfqnl_config_mode)ACCESS_ONCE(queue->copy_mode)) {
 	case NFQNL_COPY_META:
 	case NFQNL_COPY_NONE:
@@ -436,6 +456,16 @@
 		}
 #endif
 	}
+#ifdef CONFIG_NFNETLINK_COPY_UNTANGLE_MODE
+	if (queue->untangle_mode){
+		ct = nf_ct_get(entskb, &ctinfo);
+		if ( ct ) {
+			NLA_PUT(skb, NFQA_CTINFO, sizeof(ctinfo), &ctinfo);
+			NLA_PUT(skb, NFQA_CT_DIR_ORIGINAL, sizeof(ct->tuplehash[IP_CT_DIR_ORIGINAL].tuple),  &ct->tuplehash[IP_CT_DIR_ORIGINAL].tuple);
+			NLA_PUT(skb, NFQA_CT_DIR_REPLY, sizeof(ct->tuplehash[IP_CT_DIR_REPLY].tuple), &ct->tuplehash[IP_CT_DIR_REPLY].tuple);
+	  	}
+	}
+#endif
 
 	if (entskb->mark &&
 	    nla_put_be32(skb, NFQA_MARK, htonl(entskb->mark)))
@@ -722,6 +752,12 @@
 	int status = 0;
 
 	spin_lock_bh(&queue->lock);
+#ifdef CONFIG_NFNETLINK_COPY_UNTANGLE_MODE
+	if (mode & NFQNL_COPY_UNTANGLE_MODE) {
+	        queue->untangle_mode = 1;
+		mode = mode & ~NFQNL_COPY_UNTANGLE_MODE;
+	}
+#endif
 	switch (mode) {
 	case NFQNL_COPY_NONE:
 	case NFQNL_COPY_META:
