diff -urN linux-3.10.49.orig/arch/arm/mach-brcm-hnd/gpio.c linux-3.10.49.gpio/arch/arm/mach-brcm-hnd/gpio.c
--- linux-3.10.49.orig/arch/arm/mach-brcm-hnd/gpio.c	2014-08-24 09:11:52.000000000 -0700
+++ linux-3.10.49.gpio/arch/arm/mach-brcm-hnd/gpio.c	2014-08-24 09:15:01.000000000 -0700
@@ -159,6 +159,38 @@
 int isdefault = 0;
 static struct class *gpio_class = NULL;
 
+
+static void set_gpio_pin(int pin, int value)
+{
+	int val;
+	val = si_gpioouten(gpio_sih, 0, 0, GPIO_APP_PRIORITY);  
+	val |= 1 << pin;
+	si_gpioouten(gpio_sih, ~0, val, GPIO_HI_PRIORITY);
+	        
+	val = si_gpioout(gpio_sih, 0, 0, GPIO_APP_PRIORITY);
+	if (value) {
+		val |= (1 << pin);
+	} else {
+		val &= ~(1 << pin);
+	}
+	si_gpioout(gpio_sih, ~0, val, GPIO_HI_PRIORITY);
+	return;
+}
+
+
+static int get_gpio_pin(int pin)
+{
+	int val;
+	val = si_gpioouten(gpio_sih, 0, 0, GPIO_APP_PRIORITY);  
+	val &= ~(1 << pin);
+	si_gpioouten(gpio_sih, ~0, val, GPIO_HI_PRIORITY);
+
+	val = si_gpioin(gpio_sih);
+	val = (val & (1 << pin)) ? 1 : 0;
+	        
+	return val;
+}
+
 static int __init gpio_init(void)
 {
 	int i;
@@ -188,6 +220,14 @@
 		isbuffalo = 1;
 		set_hc595_reset();
 		gpio_init_flag = 1;
+
+		/*disable pin 10 and enable pin 9*/
+		printk(KERN_EMERG "GPIO PIN 9 orig value=%d\n", get_gpio_pin(9));
+		printk(KERN_EMERG "GPIO PIN 10 orig value=%d\n", get_gpio_pin(10));
+		set_gpio_pin(10,0);
+		set_gpio_pin(9,1);
+		printk(KERN_EMERG "GPIO PIN 9 value=%d\n", get_gpio_pin(9));
+		printk(KERN_EMERG "GPIO PIN 10 value=%d\n", get_gpio_pin(10));
 		return 0;
 
 	}
