 Makefile |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.16-ck11/Makefile
===================================================================
--- linux-2.6.16-ck11.orig/Makefile	2006-05-21 14:39:35.000000000 +1000
+++ linux-2.6.16-ck11/Makefile	2006-05-21 14:39:47.000000000 +1000
@@ -1,7 +1,7 @@
 VERSION = 2
 PATCHLEVEL = 6
 SUBLEVEL = 16
-EXTRAVERSION = .17
+EXTRAVERSION = -cks11
 NAME=Sliding Snow Leopard
 
 # *DOCUMENTATION*
