 block/Kconfig.iosched |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.16-ck9/block/Kconfig.iosched
===================================================================
--- linux-2.6.16-ck9.orig/block/Kconfig.iosched	2006-05-02 10:32:39.000000000 +1000
+++ linux-2.6.16-ck9/block/Kconfig.iosched	2006-05-02 10:39:52.000000000 +1000
@@ -40,7 +40,7 @@ config IOSCHED_CFQ
 
 choice
 	prompt "Default I/O scheduler"
-	default DEFAULT_AS
+	default DEFAULT_DEADLINE
 	help
 	  Select the I/O scheduler which will be used by default for all
 	  block devices.
