 mm/vmscan.c |    2 ++
 1 files changed, 2 insertions(+)

Index: linux-2.6.16-ck11/mm/vmscan.c
===================================================================
--- linux-2.6.16-ck11.orig/mm/vmscan.c	2006-05-21 12:20:12.000000000 +1000
+++ linux-2.6.16-ck11/mm/vmscan.c	2006-05-21 12:20:13.000000000 +1000
@@ -1396,6 +1396,8 @@ static int effective_sc_prio(struct task
 	if (likely(p->mm)) {
 		if (rt_task(p))
 			return -20;
+		if (idleprio_task(p))
+			return 19;
 		return task_nice(p);
 	}
 	return 0;
