 include/linux/ioprio.h |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

Index: linux-2.6.16-ck11/include/linux/ioprio.h
===================================================================
--- linux-2.6.16-ck11.orig/include/linux/ioprio.h	2006-05-21 12:20:06.000000000 +1000
+++ linux-2.6.16-ck11/include/linux/ioprio.h	2006-05-21 12:20:07.000000000 +1000
@@ -57,7 +57,7 @@ static inline int task_nice_ioprio(struc
 
 	if (idleprio_task(task))
 		effective_nice = 19;
-	else if (iso_task(task))
+	else if (rt_task(task) || iso_task(task))
 		effective_nice = -20;
 	return (effective_nice + 20) / 5;
 }
