--- linux-4.4/drivers/net/wireless/Kconfig	2016-03-03 19:36:44.000000000 +0100
+++ linux-4.4/drivers/net/wireless/Kconfig.bak	2016-11-23 12:03:37.413682908 +0100
@@ -284,16 +284,16 @@
 source "drivers/net/wireless/zd1211rw/Kconfig"
 source "drivers/net/wireless/mwifiex/Kconfig"
 source "drivers/net/wireless/cw1200/Kconfig"
-if RALINK_DEVICE
-source "drivers/net/wireless/rt3352/rt2860v2_ap/Kconfig"
-source "drivers/net/wireless/rt3352/rt2860v2_sta/Kconfig"
-endif
-if SOC_MT7620_OPENWRT
-source "drivers/net/wireless/rt7620/rt2860v2_ap/Kconfig"
-source "drivers/net/wireless/rt7620/rt2860v2_sta/Kconfig"
-source "drivers/net/wireless/rt5592/Kconfig"
-source "drivers/net/wireless/rt7612/rlt_wifi/Kconfig"
-endif
+# if RALINK_DEVICE
+# source "drivers/net/wireless/rt3352/rt2860v2_ap/Kconfig"
+# source "drivers/net/wireless/rt3352/rt2860v2_sta/Kconfig"
+# endif
+# if SOC_MT7620_OPENWRT
+# source "drivers/net/wireless/rt7620/rt2860v2_ap/Kconfig"
+# source "drivers/net/wireless/rt7620/rt2860v2_sta/Kconfig"
+# source "drivers/net/wireless/rt5592/Kconfig"
+# source "drivers/net/wireless/rt7612/rlt_wifi/Kconfig"
+# endif
 source "drivers/net/wireless/rsi/Kconfig"
 
 endif # WLAN
