--- v3.14.orig/net/ipv4/tcp_ipv4.c	2015-11-15 18:03:19.284834087 -0700
+++ v3.14/net/ipv4/tcp_ipv4.c	2015-11-15 17:49:12.610789571 -0700
@@ -1987,6 +1987,7 @@
 		goto discard_and_relse;
 	}
 
+        if (!(skb->mark&0x10000000))
 	if (!xfrm4_policy_check(sk, XFRM_POLICY_IN, skb))
 		goto discard_and_relse;
 	nf_reset(skb);
@@ -2025,6 +2026,7 @@
 	return ret;
 
 no_tcp_socket:
+        if (!(skb->mark&0x10000000))
 	if (!xfrm4_policy_check(NULL, XFRM_POLICY_IN, skb))
 		goto discard_it;
 
@@ -2047,6 +2049,7 @@
 	goto discard_it;
 
 do_time_wait:
+        if (!(skb->mark&0x10000000))
 	if (!xfrm4_policy_check(NULL, XFRM_POLICY_IN, skb)) {
 		inet_twsk_put(inet_twsk(sk));
 		goto discard_it;
