diff -ruN v3.14.orig/net/netfilter/xt_socket.c v3.14/net/netfilter/xt_socket.c
--- v3.14.orig/net/netfilter/xt_socket.c	2015-11-15 18:03:19.272833996 -0700
+++ v3.14/net/netfilter/xt_socket.c	2015-11-15 18:24:20.406547303 -0700
@@ -199,7 +199,7 @@
 EXPORT_SYMBOL(xt_socket_get4_sk);
 
 static bool
-socket_match(const struct sk_buff *skb, struct xt_action_param *par,
+socket_match(struct sk_buff *skb, struct xt_action_param *par,
 	     const struct xt_socket_mtinfo1 *info)
 {
 	struct sock *sk;
@@ -232,6 +232,14 @@
 			sk = NULL;
 	}
 
+    /* Untangle hack - or the socket mark to the packet if found
+     * We do this because there is no conntrack, and it is necessary
+     * to restore the QoS/bandwidth control mark on packets
+     * returning to nonlocally bound sockets */
+    if (sk != NULL) {
+        skb->mark |= sk->sk_mark;
+    }
+    
 	return (sk != NULL);
 }
 
