--- linux-3.18/drivers/net/wireless/Kconfig~	2016-08-11 14:25:18.000000000 -0700
+++ linux-3.18/drivers/net/wireless/Kconfig	2016-08-11 15:21:47.499213137 -0700
@@ -282,16 +282,16 @@
 source "drivers/net/wireless/zd1211rw/Kconfig"
 source "drivers/net/wireless/mwifiex/Kconfig"
 source "drivers/net/wireless/cw1200/Kconfig"
-if RALINK_DEVICE
-source "drivers/net/wireless/rt3352/rt2860v2_ap/Kconfig"
-source "drivers/net/wireless/rt3352/rt2860v2_sta/Kconfig"
-endif
-if SOC_MT7620_OPENWRT
-source "drivers/net/wireless/rt7620/rt2860v2_ap/Kconfig"
-source "drivers/net/wireless/rt7620/rt2860v2_sta/Kconfig"
-source "drivers/net/wireless/rt5592/Kconfig"
-source "drivers/net/wireless/rt7612/rlt_wifi/Kconfig"
-endif
+# if RALINK_DEVICE
+# source "drivers/net/wireless/rt3352/rt2860v2_ap/Kconfig"
+# source "drivers/net/wireless/rt3352/rt2860v2_sta/Kconfig"
+# endif
+# if SOC_MT7620_OPENWRT
+# source "drivers/net/wireless/rt7620/rt2860v2_ap/Kconfig"
+# source "drivers/net/wireless/rt7620/rt2860v2_sta/Kconfig"
+# source "drivers/net/wireless/rt5592/Kconfig"
+# source "drivers/net/wireless/rt7612/rlt_wifi/Kconfig"
+# endif
 source "drivers/net/wireless/rsi/Kconfig"
 
 endif # WLAN
